### PR TITLE
Fix body stage insufficient backpressure

### LINKED ISSUE #7331: Insufficient backpressure during body stage
### Describe the feature

ref https://github.com/paradigmxyz/reth/issues/7233

while we limit the size of buffered responses and slow down followup requests:

https://github.com/paradigmxyz/reth/blob/a6d4f6bd15ece660ab348f46a047ea9b47b71959/crates/net/downloaders/src/bodies/bodies.rs#L59-L60

https://github.com/paradigmxyz/reth/blob/a6d4f6bd15ece660ab348f46a047ea9b47b71959/crates/net/downloaders/src/bodies/bodies.rs#L373-L376

this does not take the already queued blocks into account:

https://github.com/paradigmxyz/reth/blob/a6d4f6bd15ece660ab348f46a047ea9b47b71959/crates/net/downloaders/src/bodies/bodies.rs#L61-L62

it's possible that this buffer runs away if there are requests that prevent us from yielding a new batch in time.

## TODO
* limit new requests if queued bodies exceed a certain threshold

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ghost**
Closes https://github.com/paradigmxyz/reth/issues/7331
@mattsse 

This will fix potential error regarding memory overflow in bodies download stages.

### CODE REVIEW

File: crates/net/downloaders/src/bodies/bodies.rs [hunk 1]
```diff
@@ -515,6 +523,7 @@ impl Default for BodiesDownloaderBuilder {
             stream_batch_size: 1_000,
             max_buffered_blocks_size_bytes: 2 * 1024 * 1024 * 1024, // ~2GB
             concurrent_requests_range: 5..=100,
+            max_queued_bodies_size: 100,
```
Comment (mattsse) [inline crates/net/downloaders/src/bodies/bodies.rs:526]: I think we want to limit this by using the targeted stream_batch_size, so we cant misconfigure this

I think a proper cutoff could be `self.queued_bodies.len() < 2 * stream_batch_size`

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/net/downloaders/src/bodies/bodies.rs (hunk 1 above)
Comment (0xqd) [reply crates/net/downloaders/src/bodies/bodies.rs:526]: good idea.

Comment (0xqd) [reply crates/net/downloaders/src/bodies/bodies.rs]: I have squashed the commits and update with the idea.

File: crates/net/downloaders/src/bodies/bodies.rs [hunk 2]
```diff
@@ -263,6 +263,13 @@ where
         }
         None
     }
+
+    /// Returns true if a new request can be submitted
+    fn can_submit_new_request(&self) -> bool {
+        self.queued_bodies.len() < 2 * self.stream_batch_size &&
```
Comment (mattsse) [inline crates/net/downloaders/src/bodies/bodies.rs:269]: we should document this step abit

**Review by mattsse — CHANGES_REQUESTED** [review]
just doc nit

pending @Rjected

**Review by Rjected — APPROVED** [review]
lgtm, pending doc nit as well

**Review by mattsse — APPROVED** [review]
lgtm,

ty!

**Pull request merged**